### PR TITLE
CIRC-1747 - Patron notified for all recalls - not only those that change the due date

### PR DESCRIPTION
**Pull request by gurleenkaurbp**
https://issues.folio.org/browse/CIRC-1747 - Patron notified for all recalls - not only those that change the due date
## Purpose

The purpose of this PR is to address the issue w.r.t a borrower receives recall notices for each request that is being placed for recall by a  requester. As per Julie and Brooks the borrower should receive the recall notice only the first time when it was recalled by a requester. The reason is because the due date will not change after the first recall request is placed therefore, there isnt a need to send recall notice emails again and again to the borrower. (spam the borrower's inbox)

## Approach

The approach taken is to check the request queue size for that item for open-not yet filled recalls request type. If the queue size is >1 then no need to publish the recall notice. Otherwise, a recall notice event will be published and email will be received by the borrower.
#### TODOS and Open Questions
- [x] Check logging

### CODE REVIEW

File: src/test/java/api/queue/RequestQueueResourceTest.java [hunk 1]
```diff
@@ -609,7 +609,7 @@ void logRecordEventIsPublished(TlrFeatureStatus tlrFeatureStatus) {
     verifyQueueUpdatedForItem(reorderQueue, response);
 
     // TODO: understand why
-    int numberOfPublishedEvents = 17;
+    int numberOfPublishedEvents = 16;
```
[the remark this answers is not in the document]
Comment (gurleenkaurbp) [reply src/test/java/api/queue/RequestQueueResourceTest.java:612]: As you can see this test is creating two recalls -- 
.addReorderRequest(firstRecallRequest.getId().toString(), 2)
      .addReorderRequest(secondRecallRequest.getId().toString(), 3)
      
      Add now for the second recall the borrower will not receive the recall notice email and that is the one less event that will not be published.

File: src/main/java/org/folio/circulation/resources/RequestNoticeSender.java [hunk 2]
```diff
@@ -92,10 +92,13 @@ public static RequestNoticeSender using(Clients clients) {
   private final EventPublisher eventPublisher;
   protected final LocationRepository locationRepository;
 
+  private Long recallRequestSize = 0l;
```
Comment (tatsiana-tarhonskaya) [inline src/main/java/org/folio/circulation/resources/RequestNoticeSender.java:95]: maybe long is sufficient?

Comment (tatsiana-tarhonskaya) [inline src/main/java/org/folio/circulation/resources/RequestNoticeSender.java:95]: can it be just long?

Comment (gurleenkaurbp) [reply src/main/java/org/folio/circulation/resources/RequestNoticeSender.java:95]: I mostly changing this to a boolean, my next committ will reflect it ..

File: src/test/java/api/requests/RequestsAPICreationTests.java [hunk 3]
```diff
@@ -3503,22 +3505,89 @@ void itemCheckOutRecallRequestCreationShouldProduceNotice() {
...
+      verifyNumberOfPublishedEvents(NOTICE_ERROR, 0);
+
+
+      // verify noticeLogContextItemLogs
+      validateNoticeLogContextItem(noticeLogContextItemLogs.get(0), item);
+      validateNoticeLogContextItem(noticeLogContextItemLogs.get(1), item);
+    }
   }
 
+  private Integer verifyNumberOfSentNoticesForBorrower(List<JsonObject> noticeLogContextItemLogs, IndividualResource borrower) {
```
[the remark this answers is not in the document]
Comment (gurleenkaurbp) [reply src/test/java/api/requests/RequestsAPICreationTests.java]: Renamed the method.

File: src/main/java/org/folio/circulation/resources/RequestNoticeSender.java [hunk 4]
```diff
@@ -92,10 +92,15 @@ public static RequestNoticeSender using(Clients clients) {
...
 
+  private long recallRequestSize = 0l;
+
   public Result<RequestAndRelatedRecords> sendNoticeOnRequestCreated(
     RequestAndRelatedRecords records) {
 
     Request request = records.getRequest();
+    recallRequestSize = records.getRequestQueue().getRequests()
+      .stream()
+      .filter(r -> r.getRequestType() == RequestType.RECALL && r.isNotYetFilled()).count();
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/circulation/resources/RequestNoticeSender.java:103]:
```suggestion
      .filter(r -> r.getRequestType() == RequestType.RECALL && r.isNotYetFilled())
      .count();
```

File: src/main/java/org/folio/circulation/resources/RequestNoticeSender.java [hunk 5]
```diff
@@ -92,10 +92,15 @@ public static RequestNoticeSender using(Clients clients) {
   private final EventPublisher eventPublisher;
   protected final LocationRepository locationRepository;
 
+  private long recallRequestSize = 0l;
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/circulation/resources/RequestNoticeSender.java:95]:
```suggestion
  private long recallRequestCount = 0l;
```

File: src/test/java/api/support/fakes/FakePubSub.java [hunk 6]
```diff
@@ -9,6 +9,7 @@
 
 import java.util.ArrayList;
 import java.util.Arrays;
+import java.util.Collections;
```
Comment (OleksandrVidinieiev) [inline src/test/java/api/support/fakes/FakePubSub.java:12]: Redundant import

File: src/test/java/api/requests/RequestsAPICreationTests.java [hunk 7]
```diff
@@ -3503,22 +3505,89 @@ void itemCheckOutRecallRequestCreationShouldProduceNotice() {
...
+    requestsFixture.placeItemLevelHoldShelfRequest(item, requester, requestDate, "Recall");
+
+    List<JsonObject> noticeLogContextItemLogs = Awaitility.waitAtMost(1, TimeUnit.SECONDS)
+      .until(() -> getPublishedEventsAsList(byLogEventType(NOTICE)), hasSize(4));
+      // notice for the recall is expected
+    verifyNumberOfSentNotices(4);
+    assertEquals(2, getNumberOfRecallNoticesForBorrower(noticeLogContextItemLogs, borrower));
+    verifyNumberOfPublishedEvents(NOTICE, 4);
+    verifyNumberOfPublishedEvents(NOTICE_ERROR, 0);
+
```
Comment (OleksandrVidinieiev) [inline src/test/java/api/requests/RequestsAPICreationTests.java:3572]: Redundant empty line

File: src/test/java/api/requests/RequestsAPICreationTests.java [hunk 8]
```diff
@@ -3503,22 +3505,89 @@ void itemCheckOutRecallRequestCreationShouldProduceNotice() {
...
+
+      // verify noticeLogContextItemLogs
```
Comment (OleksandrVidinieiev) [inline src/test/java/api/requests/RequestsAPICreationTests.java:3574]:
```suggestion
    // verify noticeLogContextItemLogs
```

File: src/test/java/api/requests/RequestsAPICreationTests.java [hunk 9]
```diff
@@ -3503,22 +3505,89 @@ void itemCheckOutRecallRequestCreationShouldProduceNotice() {
...
     requestsFixture.placeItemLevelHoldShelfRequest(item, requester, requestDate, "Recall");
+    requestsFixture.placeItemLevelHoldShelfRequest(item, requester2, requestDate, "Recall");
 
-    // notice for the recall is expected
-    verifyNumberOfSentNotices(2);
-    verifyNumberOfPublishedEvents(NOTICE, 2);
+    List<JsonObject> noticeLogContextItemLogs = Awaitility.waitAtMost(1, TimeUnit.SECONDS)
+      .until(() -> getPublishedEventsAsList(byLogEventType(NOTICE)), hasSize(3));
+
+      // notice for the recall is expected
```
Comment (OleksandrVidinieiev) [inline src/test/java/api/requests/RequestsAPICreationTests.java:3519]:
```suggestion
    // notice for the recall is expected
```

File: src/test/java/api/requests/RequestsAPICreationTests.java [hunk 10]
```diff
@@ -3503,22 +3505,89 @@ void itemCheckOutRecallRequestCreationShouldProduceNotice() {
...
+    verifyNumberOfSentNotices(4);
+    assertEquals(2, getNumberOfRecallNoticesForBorrower(noticeLogContextItemLogs, borrower));
+    verifyNumberOfPublishedEvents(NOTICE, 4);
+    verifyNumberOfPublishedEvents(NOTICE_ERROR, 0);
+
+
+      // verify noticeLogContextItemLogs
+    validateNoticeLogContextItem(noticeLogContextItemLogs.get(0), item);
+    validateNoticeLogContextItem(noticeLogContextItemLogs.get(1), item);
+
```
Comment (OleksandrVidinieiev) [inline src/test/java/api/requests/RequestsAPICreationTests.java:3577]: Redundant empty line

File: src/test/java/api/requests/RequestsAPICreationTests.java [hunk 11]
```diff
@@ -3503,22 +3505,89 @@ void itemCheckOutRecallRequestCreationShouldProduceNotice() {
...
+  }
+
+  private Integer getNumberOfRecallNoticesForBorrower(List<JsonObject> noticeLogContextItemLogs, IndividualResource borrower) {
```
Comment (OleksandrVidinieiev) [inline src/test/java/api/requests/RequestsAPICreationTests.java:3580]: Few thoughts on this method:
- recall is mentioned in method's name but not in the body
- name suggests that it's counting patron notices, while in fact it is counting notice events (not the same thing)
- you call this method twice and both times use its return value immediately to compare it against expected value, so maybe this method should do the verification too?

```
  private void verifyNumberOfNoticeEventsForUser(UUID userId, int expectedNoticeEventsCount) {
    long noticeEventsCount = getPublishedEventsAsList(byLogEventType(NOTICE))
      .stream()
      .filter(event ->
        new JsonObject(event.getString("eventPayload"))
          .getJsonObject("payload")
          .getString("userId")
          .equals(userId.toString()))
      .count();

    assertThat(noticeEventsCount, is(expectedNoticeEventsCount));
  }

```

File: src/test/java/api/requests/RequestsAPICreationTests.java [hunk 12]
```diff
@@ -3503,22 +3505,80 @@ void itemCheckOutRecallRequestCreationShouldProduceNotice() {
...
+    verifyNumberOfSentNotices(4);
+    verifyNumberOfNoticeEventsForUser(requester.getId(), 2l);
+    verifyNumberOfPublishedEvents(NOTICE, 4);
+    verifyNumberOfPublishedEvents(NOTICE_ERROR, 0);
     validateNoticeLogContextItem(noticeLogContextItemLogs.get(0), item);
     validateNoticeLogContextItem(noticeLogContextItemLogs.get(1), item);
   }
 
+  private void verifyNumberOfNoticeEventsForUser(UUID userId, long expectedNoticeEventsCount) {
+    long noticeEventsCount = getPublishedEventsAsList(byLogEventType(NOTICE))
```
Comment (OleksandrVidinieiev) [inline src/test/java/api/requests/RequestsAPICreationTests.java:3570]:
```suggestion
  private void verifyNumberOfNoticeEventsForUser(UUID userId, int expectedNoticeEventsCount) {
    int noticeEventsCount = (int) getPublishedEventsAsList(byLogEventType(NOTICE))
```

**Review by OleksandrVidinieiev — APPROVED** [review]

**Review by Vignesh-kalyanasundaram — APPROVED** [review]

**Review by Arghya22 — APPROVED** [review]

**Pull request merged**